### PR TITLE
[tune] Fix `Analysis.dataframe()` documentation and enable passing of `mode=None`

### LINKED ISSUE #17923: [tune] Different in performance and best model between analysis.best_config and the best model in analysis.dataframe
### What is the problem?

*pytorch_lightning: '1.4.2'*
*ray: '1.5.2'*
*tensorboardX: '2.4'*

callbacks = TuneReportCheckpointCallback
scheduler = ASHAScheduler
search_alg = HyperOptSearch
reporter = CLIReporter

at the end of the tune.run (searching for the best performance on the validation loss) i obtain the best model (best config) and the best checkpoint.
When i move the result in a df using analysis.dataframe if i search for the minimum value for the validation loss, frequently i obtain a different configuration with respect to the best model.
the pl model it runs with earlystopping on validation loss

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

cc @amogkam 

## Related issue number

Closes #17923

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/analysis/experiment_analysis.py [hunk 1]
```diff
@@ -343,6 +355,7 @@ def _retrieve_rows(self,
                        metric: Optional[str] = None,
                        mode: Optional[str] = None) -> Dict[str, Any]:
         assert mode is None or mode in ["max", "min"]
+        assert not mode or metric
```
Comment (Yard1) [inline python/ray/tune/analysis/experiment_analysis.py:358]: Is this supposed to be `!mode or metric` or `!(mode or metric)`? Could we use parenthesis here to make it clearer?

**Review by Yard1 — APPROVED** [review]
LGTM, just one question!

File: python/ray/tune/analysis/experiment_analysis.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/analysis/experiment_analysis.py:358]: It is supposed to be `!mode or metric` i.e. if mode is set, metric has to be set as well
